R/epiparameter.R

```diff
@@ -0,0 +1,996 @@
+#' @export
+generate.epiparameter <- function(x, times, ...) {
+  if (isFALSE(is_parameterised(x))) {
+    stop("<epiparameter> is unparameterised", call. = FALSE)
+  }
+  # check times is a single number for consistent behaviour
+  checkmate::assert_number(times)
+  if (inherits(x$prob_distribution, "distcrete")) {
+    unlist <- length(x$prob_distribution) == 1
+    out <- x$prob_distribution$r(n = times)
+    out <- if (unlist) unlist(out, recursive = FALSE) else out
+  } else {
+    recursive <- length(x$prob_distribution) == 1
+    out <- distributional::generate(x$prob_distribution, times = times)
+    out <- unlist(out, recursive = recursive)
+  }
+  out
+}
+
+#' Discretises a continuous distribution in an `<epiparameter>` object
+#'
+#' @details Converts the S3 distribution object in an `<epiparameter>` from
+#' continuous (using an object from the `{distributional}` package) to a
+#' discretised distribution (using an object from the `{distcrete}` package).
+#'
+#' @inheritParams print.epiparameter
+#' @param ... [dots] Extra arguments to be passed to the method.
+#'
+#' @inherit epiparameter return
+#' @export
+#'
+#' @examples
+#' ebola_incubation <- epiparameter(
+#'   disease = "ebola",
+#'   epi_name = "incubation_period",
+#'   prob_distribution = create_prob_distribution(
+#'     prob_distribution = "gamma",
+#'     prob_distribution_params = c(shape = 1, scale = 1)
+#'   )
+#' )
+#' discretise(ebola_incubation)
+discretise <- function(x, ...) {
+  UseMethod("discretise")
+}
+
+#' @rdname discretise
+#' @export
+discretize <- discretise
+
+#' @rdname discretise
+#' @export
+discretise.epiparameter <- function(x, ...) {
+  # check if distribution is already discretised if so return early
+  if (inherits(x$prob_distribution, "distcrete")) {
+    message("Distribution in `epiparameter` is already discretised")
+    return(x)
+  } else {
+    # extract prob dist and prob dist parameters from epiparameter
+    prob_dist <- family(x)
+    if (prob_dist == "mixture") {
+      warning(
+        "<epiparameter> object with mixture distributions cannot be ",
+        "discretised.\n Returning undiscretised <epiparameter>.",
+        call. = FALSE
+      )
+      return(x)
+    }
+    prob_dist_params <- get_parameters(x)
+
+    # if distribution is truncated take only parameters
+    if (is_truncated(x)) {
+      warning(
+        "Discretising a truncated continuous distribution, ",
+        "returning non-truncated discretised distribution",
+        call. = FALSE
+      )
+
+      idx <- grep(
+        pattern = "lower|upper",
+        x = names(prob_dist_params),
+        ignore.case = TRUE
+      )
+      prob_dist_params <- prob_dist_params[-idx]
+
+      # trunc dist family is truncated so get prob dist by unclassing dist and
+      # extracting name
+      list_dist <- unclass(x$prob_distribution)
+      prob_dist <- gsub(
+        pattern = "dist_",
+        replacement = "",
+        x = class(list_dist[[1]][[1]])[1],
```

Comment:
Might it be helpful to add a comment explaining the indexing here? This is convoluted to read and some assistance would help (me in any case).

Reply:
This has been resolved in PR #398. This PR generalises `.get_mixture_dist()` to `.distributional_family()` which is now called here instead of the multiple consecutive subsetting.